lib/kernel/src/inet_dns_tsig.erl

```diff
@@ -0,0 +1,340 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2023. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(inet_dns_tsig).
+
+%% DNS TSIG handling
+%%
+%% RFC 8945: Secret Key Transaction Authentication for DNS (TSIG)
+
+%% WARNING: in the spirit of inet_dns.erl, this module only handles the
+%% cryptographic operations and does not absolve you of your responsibities
+%% that include:
+%%  * implement a truncation policy as per RFC8945, section 5.2
+%%  * verifying that for TSIG over TCP the last message contains a TSIG RR
+%%  * truncated messages must include the TSIG RR as per RFC8945, section 5.3
+%%  * if running as a server, generate suitable error responses
+%%    as per RFC8945, section 5.2
+
+-export([init/1, sign/2, sign/3, verify/3]).
+
+-include("inet_dns.hrl").
+
+-define(ALGS_SUPPORTED, [md5, sha, sha224, sha256, sha384, sha512, sha512]).
+-define(ALG_DEFAULT, sha256). % RFC8945, section 6
+
+-define(FUDGE_DEFAULT, 300). % RFC8945, section 10
+
+-define(MAC_SIZE_MIN, 10). % RFC8945, section 5.2.2.1
+-define(TTL, 0).
+
+-define(NOW, os:system_time(seconds)). %  RFC8945, section 5.2.3
+
+-type tsig_key() :: {string(),iodata()}.
+
+-type tsig_algs() :: crypto:hmac_hash_algorithm() | gss_tsig.
+-type tsig_alg() :: tsig_algs() | {tsig_algs(),?MAC_SIZE_MIN..65535}.
+
+-type tsig_error() :: ?BADSIG | ?BADKEY | ?BADTIME | ?BADTRUNC.
+
+-record(tsig_state, {
+    key       :: tsig_key() | list(tsig_key()),
+    alg       :: tsig_alg(), % client only
+    fudge     :: 0..65535,
+
+    qr    = 0 :: 0..100, % 0 (false): request, 1+ (true): response
+                         % this value is overloaded to handle TSIG
+                         % over TCP where we need to accept up to
+                         % 99 messages without TSIG so each time a
+                         % response with a TSIG RR is received this
+                         % value is reset to 2
+    id        :: undefined | 0..65535,
+    mac       :: undefined | binary() | crypto:mac_state()
```

Comment:
Since `crypto:mac_state()` is an opaque type it cannot be assumed to not be e.g `undefined`, so when comparing if this field is `undefined` you break the opaque type specification.  It cannot be compared.

Therefore a field with an opaque type has to have a different field (or other way of knowing) that flags if it contains a `crypto:mac_state()` or not.

Reply:
What about instead something like?
```
mac :: {?MODULE,undefined | binary()} | {crypto,crypto:mac_state()}
```

Partitions off "this is mine" and "this is yours".

If you prefer a separate element in the record, happy to do that instead?

Reply:
That would be just fine, it does not have to be a separate element

Reply:
Resolved by 5c1728c8591a398683d64918ff9127e4c44bada6